DM-15465: Clean up utils usage with a view to supporting test running in pull requests

timj opened the pull request:
* Removed getPackageDir and replaced with package resources
* Remove spurious usage of lsst.utils.tests in one test file

Still have many doImport usages and one test uses a getTempFilePath call.

timj committed:
Allow ButlerURI to be created from another ButlerURI

timj committed:
Switch ButlerConfig.configDir to be a ButlerURI

Make use of new APIs in ButlerURI

timj committed:
Switch config discovery from getPackageDir to pkg_resources

This required that the configs move inside the python package.
It also needed some cleanup in ButlerURI versus bare paths
and the addition of support to instantiate a Config from
a new Resource class.

timj committed:
Remove lsst.utils from test_dbAuth

No other daf_butler tests include the memory tester
so the usage here was anomalous.

timj committed:
Add the __init__.py back at top level

Python really really wants it there and setuptools won't
work without it. It already caused problems with mypy
so put it back.

timj committed:
Create a setup.py to enable test builds with pip

If lsst.utils is not available locally the files we need
are downloaded.

timj committed:
Add an action to build the package and run tests

timj committed:
Explicitly preinstall sphgeom dependencies

Not sure why the next line doesn't do this

timj committed:
Allow longer lines in YAML files

timj committed:
Fall back to standard safe loader if C yaml library is not found

timj committed:
Ignore files left over from python setup.py build

timj committed:
Add psutil, needed by lsst.utils.tests

timj committed:
Force installation of sqlite and libyaml-dev

timj committed:
Use ubuntu 20.04 for modern sqlite

timj committed:
Adjust tests for missing arguments to account for click 7.1 change

click 7.0 uses double quotes but 7.1 uses a single quote.
For now we support both so switch the code to a regex rather
than a substring match.

timj committed:
Install pytest-xdist in workflow to speed up tests

timj committed:
Enable doc build in action

timj commented:
This now seems to be passing so that pull requests will run the tests and build the documentation.  It takes 8minutes to run and is currently not required for merging. Two minutes of that is building sphgeom so maybe we can upload a wheel as a "GitHub release" and pull that to save some time.

I wonder if we have to reconsider the pull_request and push builds. Are they the same once a pull request has been created?

TallJimbo commented on python/lsst/daf/butler/tests/cliCmdTestBase.py:
```diff
@@ -103,11 +103,11 @@ def run_missing(self, inputs, expectedMsg):
         result = self.run_command(inputs)
         self.assertNotEqual(result.exit_code, 0, clickResultMsg(result))
-        self.assertIn(expectedMsg, result.stdout)
+        self.assertRegex(result.stdout, expectedMsg)
```
If these tests cause problems ever again I vote for dropping them.  I don't think I've ever met a stringification test that added net value, even for classes that actually had broken stringification.

TallJimbo reviewed (approved):
Only big picture comment is that with all of the conditionals in `ButlerURI` I wonder if it'd work better as an ABC with separate implementations for remote URIs, POSIX paths, and now pkg_resources.  Or maybe a concrete class that delegated to an ABC or something.  Probably out of scope for this ticket, but something to think about.

timj committed:
Fix documentation references to DAF_BUTLER_DIR

timj committed:
Change sphgeom URLs to point to master

timj commented:
Regarding ButlerURI returning subclasses for different URI schemes, that does seem worth thinking about. It would also allows the addition of more concrete methods such as "does this item exist" and "download this resource".

ktlim commented on doc/lsst.daf.butler/configuring.rst:
```diff
@@ -22,7 +22,7 @@ To construct a Butler configuration object (`~lsst.daf.butler.ButlerConfig`) fro
   The contents of these files will then be inserted into the configuration at the same hierarchy as the ``includeConfigs`` directive, with priority given to the values defined explicitly in the parent configuration (for lists of include files later files overwrite content from earlier ones).
 * Each sub configuration class is constructed by supplying the relevant subset of the global config to the component Config constructor.
-* A search path is constructed by concatenating the supplied search path, the environment variable path (``$DAF_BUTLER_CONFIG_PATH``), and the daf_butler config directory (``$DAF_BUTLER_DIR/config``).
+* A search path is constructed by concatenating the supplied search path, the environment variable path (``$DAF_BUTLER_CONFIG_PATH``), and the daf_butler config directory (using the ``lsst.daf.butler`` package resource).
```
```suggestion
* A search path is constructed by concatenating the supplied search path, the environment variable path (``$DAF_BUTLER_CONFIG_PATH``), and the daf_butler config directory (within the ``lsst.daf.butler`` package resource).
```

ktlim commented on .github/workflows/build.yaml:
```diff
@@ -0,0 +1,47 @@
+      - name: Run tests
+        run: pytest -r a -v -n 3
+
+      - name: Install documenteer
```
Note for the future: Should package the install and build as an action that others can reuse.

ktlim commented on .github/workflows/build.yaml:
```diff
@@ -0,0 +1,47 @@
+      - name: Install sphgeom dependencies
+        run: pip install -r https://raw.githubusercontent.com/lsst/sphgeom/master/requirements.txt
+
+      - name: Install dependencies
```
Note for the future: these four steps (deps, xdist, install ., pytest) could be packaged as an action for others to reuse.

timj replied:
An action can import another action?

ktlim replied:
A _workflow_ (which this is) uses actions.  I'm saying to package up the `run` scripts as a reusable action.

ktlim commented on setup.cfg:
```diff
@@ -1,3 +1,50 @@
+name = daf_butler
+description = An abstraction layer for reading and writing astronomical data to datastores
+author = Rubin Observatory Data Management
+author_email = support@lsst.org
```
Who does this go to?

timj replied:
No idea. It was a placeholder. Since I'm not uploading to PyPI I guess it's better to remove it.

ktlim commented on setup.cfg:
```diff
@@ -1,3 +1,50 @@
+description = An abstraction layer for reading and writing astronomical data to datastores
+author = Rubin Observatory Data Management
+author_email = support@lsst.org
+license = GPLv3+
```
I thought this was discouraged in favor of `classifiers`: https://packaging.python.org/guides/distributing-packages-using-setuptools/#license

timj replied:
Probably. I am using the classifiers so I'll remove this. It was mostly historical copy and paste.

ktlim commented on tests/test_dbAuth.py:
```diff
@@ -420,14 +419,5 @@ def test_urlEncoding(self):
             "%2A%5B%5D@host.example.com:5432/my_database")
 
 
-class MemoryTester(lsst.utils.tests.MemoryTestCase):
```
We're losing the open-fd test by doing this, right?

timj replied:
Yes. We weren't using it anywhere else though. I had been thinking of using pytest-openfiles instead but I've never been able to convince pytest-openfiles that there was a leak when I experimented with it in the past.

timj replied:
I've added pytest-openfiles and it found a real problem: our "test repo" tests that were meant to be using in-memory registry for speed were in fact using sqlite files because we had the wrong argument for makeRepo. We should add it to the conda env.

ktlim commented on requirements.txt:
```diff
@@ -0,0 +1,16 @@
+pyyaml >= 5.1
```
Unfortunate that we need these twice: here and in `setup.cfg`.  Is this the standard way to maintain these?

timj replied:
I will admit that @jonathansick tried to explain to me once why we have to duplicate requirements but I haven't really grokked it yet. You can't do "pip install requirements from setup.cfg" so a separate file gives you that.  setuptools doesn't know about requirements.txt so ignores them and uses the ones from setup.cfg.

jonathansick replied:
I don't understand the strategy here. If you've got a `setup.cfg`, then you shouldn't need this `requirements.txt` file at all. What is the `requirements.txt` file being used for?

timj replied:
Ok. I thought we had a whole conversation about this back when you were here.  I was following: https://packaging.python.org/discussions/install-requires-vs-requirements/  to a certain extent except I wasn't really doing the "minimal in setup.cfg and maximal in requirements.txt approach".

I'm pretty sure you can't ask pip to install requirements specified in a setup.cfg so you absolutely need the requirements.txt for that (and you need to somehow install setuptools_cpp when building sphgeom before you even run setup.py since that's required to configure the C++ build). On the other hand setuptools doesn't look at requirements.txt and I don't see situations where people try to read the requirements.txt into their setup.py execution. Also requirements.txt lets people preinstall dependencies using pip before they try to install the package itself.  It seems like an inconsistent mess at first glance.

ktlim commented on python/lsst/daf/butler/core/config.py:
```diff
@@ -25,11 +25,14 @@
 import pprint
 import os
+import pkg_resources
+import posixpath
```
You should add a comment to explain why this is preferred over `os.path` and `pathlib`.

ktlim commented on python/lsst/daf/butler/core/config.py:
```diff
@@ -353,7 +456,13 @@ def _processExplicitIncludes(self):
         searchPaths = [os.path.curdir]
         if self.configFile is not None:
-            searchPaths.append(os.path.abspath(os.path.dirname(self.configFile)))
+            if isinstance(self.configFile, str):
```
If you could turn `str` into `ButlerURI` earlier you could get rid of this case and the one below.  But if you can turn both into a `Resource` that would be even better.

timj replied:
I can't turn an S3 URI into a `Resource` but I should be able to convert all the plain strings into schemeless URIs.  I had thought about adding a pkg_resources URI scheme to ButlerURI but ButlerURI can't do concrete things yet (like check for existence).

ktlim replied:
I think this is the same as Jim's ABC comment above.  We're not talking about `Resource` as currently defined but a new `Resource` ABC.

ktlim commented on python/lsst/daf/butler/core/config.py:
```diff
@@ -385,10 +494,37 @@ def _processExplicitIncludes(self):
+                                        found = os.path.normpath(os.path.abspath(filePath))
+                                        break
+                                else:
+                                    # S3 is harder because we can't tell if
```
You can use `head_object()` to determine if it is there by checking its metadata.

timj replied:
Or use `s3utils.s3CheckFileExists()` I think.   None of the S3 support is tested at the moment. I think I'd need to add a s3config test that uploads some test configurations to the fake moto bucket and then tries to reconstruct the config from it.  I don't think @DinoBektesevic or @hsinfang are making use of DAF_BUTLER_CONFIG_PATH at the moment or using `!include` so I'm suggesting I defer implementation until we need it.

ktlim commented on python/lsst/daf/butler/core/config.py:
```diff
@@ -140,6 +148,66 @@ def extractFile(self, filename):
 
 
+@dataclass
+class Resource:
```
Is there a way to unify filesystem paths or even URIs with these `Resource` paths so that they appear to the rest of the code as a uniform class instead of having to type-case everywhere?

timj replied:
Adding `pkg_resources://lsst.daf.butler/configs/datastore.yaml` URI support to ButlerURI might work but really needs `ButlerURI` to be updated to have concrete methods for checking existence and even opening a stream.  I think some of this ties in with @TallJimbo wondering if ButlerURI could be changed so that the constructor returns specialist S3ButlerURI, PosixFileURI and LocalFileURI implementations.

ktlim commented on setup.py:
```diff
@@ -0,0 +1,70 @@
+__all__ = ("__version__", )
+__version__='{version}'""", file=f)
+
+# The purpose of this setup.py is to build enough of the system to
```
Almost worth including it as a git submodule...

timj replied:
but then I'd still have to write a new `__init__.py` that didn't try to import all the C++ routines. I imagine that these github.com gets should be fairly stable for actions that themselves run on github.com.

ktlim commented on setup.py:
```diff
@@ -0,0 +1,70 @@
+            with open(init_path, "w") as fh:
+                print("from .doImport import *", file=fh)
+
+# read the contents of your README file
```
```suggestion
# read the contents of our README file
```

timj committed:
Fix flake8 problem with blank line

timj committed:
Update setup.py

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

timj committed:
Update doc/lsst.daf.butler/configuring.rst

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

timj committed:
Remove support email and license field from setup.cfg

timj committed:
Add comment on usage of posixpath

timj committed:
Expand on S3 not supporting includeConfigs

timj committed:
Simplify includeConfig search path by using ButlerURI

This saves a special "str" logic path.

timj committed:
Configure test repo to properly use in-memory registry

We were actually losing our override and getting a default
sqlite registry file.

timj committed:
Enable pytest openfiles checking

Disable .ttf files opened by matplotlib.
